src/Inline/Element/AbstractInline.php

```diff
@@ -26,38 +26,11 @@
     public $data = [];
 
     /**
-     * @var AbstractInline|null
+     * @return bool
      */
-    protected $parent;
-
-    /**
-     * @return AbstractBlock|AbstractInline|null
-     */
-    public function getParent()
+    public function isContainer()
     {
-        return $this->parent;
-    }
-
-    /**
-     * @param AbstractBlock|AbstractInline $parent
-     *
-     * @return $this
-     */
-    public function setParent($parent)
-    {
-        if (!$parent instanceof AbstractBlock && !$parent instanceof self) {
-            throw new \InvalidArgumentException(sprintf(
-                'Argument 1 passed to %s() must be an instance of %s or %s, instance of %s given',
-                __METHOD__,
-                'League\CommonMark\Block\Element\AbstractBlock',
-                'League\CommonMark\Block\Element\AbstractInline',
-                is_object($parent) ? get_class($parent) : gettype($parent)
-            ));
-        }
-
-        $this->parent = $parent;
-
-        return $this;
+        return false;
```

Comment:
Some inlines (like emphasis and links) **are** containers (https://github.com/jgm/commonmark.js/blob/0.21.0/lib/node.js#L3-L19)

I believe this `isContainer()` check is currently only used in the [`NodeWalker::next()` method](https://github.com/thephpleague/commonmark/pull/169/files#diff-0cfd46ad72d87b09256e734b343b397aR39).

Here's an idea - what if we create a `ContainerNode` which extends from `Node`, and move the following methods (and their related variables) into it:

* `appendChild()`
* `prependChild()`
* `getChildren()`
* `getFirstChild()`
* `getLastChild()`
* `detachChildren`
* `replaceChildren`

Then, in `NodeWalker`, replace this line:

~~~php
if ($entering && $current->isContainer()) {
~~~

With:

~~~php
if ($entering && $current instanceof ContainerNode) {
~~~

Therefore avoiding the need to define a separate `isContainer` method - you simply look at the object type.

What are your thoughts on that?